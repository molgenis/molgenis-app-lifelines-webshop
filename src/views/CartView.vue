<template>
  <div id="cart-view" class="row">
    <div class="col">
      <h3>{{$t('lifelines-webshop-cart-header')}}</h3>
      <template v-if="selectedVariableIds.length">
        <div class="mb-3" v-if="selectedVariableIds.length > 0">
          <button type="button" class="btn btn-primary save" @click="save">Save</button>
          <router-link
                class="btn btn-warning ml-2"
                type="button"
                to="/order"
                tag="button">
            Order
          </router-link>
        </div>
        <h5>Selected variables</h5>
        <spinner-animation v-show="loading" />
        <ul v-if="!loading">
          <li v-for="variableId in selectedVariableIds" :key="variableId">
            <template v-if="variablesMap[variableId].label">{{variablesMap[variableId].label}}</template>
            <template v-else>{{variablesMap[variableId].name}}</template>
            <span>{{ variableAssesments[variableId] }}</span>
          </li>
        </ul>
        <div class="mb-3" v-if="selectedVariableIds.length > 10">
          <button type="button" class="btn btn-primary save" @click="save">Save</button>
          <router-link
                class="btn btn-warning ml-2"
                type="button"
                to="/order"
                tag="button">
            Order
          </router-link>
        </div>
      </template>
      <template v-else>
        <h5>No variables selected</h5>
        <p v-if="isSignedIn">Use the shop tab to select variables to order</p>
        <p v-else>Sign in to select and order variables</p>
      </template>
    </div>
  </div>
</template>

<script>
import Vue from 'vue'
import SpinnerAnimation from '../components/animations/SpinnerAnimation.vue'
import { mapActions, mapState } from 'vuex'

export default Vue.extend({
  name: 'CartView',
  components: { SpinnerAnimation },
  methods: {
    ...mapActions(['save'])
  },
  computed: {
    ...mapState(['isSignedIn']),
    gridSelection () {
      return this.$store.state.gridSelection
    },
    selectedVariableIds () {
      return Object.keys(this.gridSelection)
    },
    variablesMap () {
      return this.$store.state.variables
    },
    assessmentsMap () {
      return this.$store.state.assessments
    },
    variableAssesments () {
      let variableAssesmentsStings = {}
      for (const [variableId, assessmentIds] of Object.entries(
        this.gridSelection
      )) {
        const assessmentNames = assessmentIds.map(
          assessmentId => this.assessmentsMap[assessmentId].name
        )
        variableAssesmentsStings[variableId] =
          '( ' + assessmentNames.join(', ') + ' )'
      }

      return variableAssesmentsStings
    },
    loading () {
      return !(
        Object.keys(this.assessmentsMap).length &&
        Object.keys(this.variablesMap).length
      )
    }
  }
})
</script>
