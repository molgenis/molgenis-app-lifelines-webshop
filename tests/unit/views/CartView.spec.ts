import { mount, createLocalVue, RouterLinkStub } from '@vue/test-utils'

import CartSection from '@/types/CartSection'
import CartView from '@/views/CartView.vue'
import Vuex from 'vuex'
import Vue from 'vue'
import flushPromises from 'flush-promises'
import BootstrapVue from 'bootstrap-vue'

Vue.filter('i18n', (value: string) => value) // Add dummy filter for i18n

describe('CartView.vue', () => {
  const localVue = createLocalVue()
  localVue.use(Vuex)
  localVue.use(BootstrapVue)

  let stubs = {
    RouterLink: RouterLinkStub,
    'toast-component': true
  }

  let mocks = {
    $router: {
      push: jest.fn()
    }
  }

  let store: any
  let actions: any
  let mutations: any

  let cartTree: any[]

  const getters = {
    cartTree: () => cartTree,
    isSignedIn: () => true,
    selectedVariableIds: () => []
  }

  beforeEach(() => {
    let state: any
    actions = {
      save: jest.fn()
    }
    mutations = {
      setToast: jest.fn(),
      removeToast: jest.fn()
    }

    state = {
      gridSelection: {
        123: [1, 2],
        456: [3]
      },
      variables: {
        123: {
          id: 123,
          name: 'var123',
          label: 'var 123',
          subvariables: [{ id: 456 }]
        },
        456: {
          id: 456,
          name: 'var456',
          label: 'var 456',
          subvariable_of: { id: 123 },
          subvariables: []
        }
      },
      assessments: {
        1: {
          id: 1,
          name: '2C'
        },
        2: {
          id: 1,
          name: '1A'
        },
        3: {
          id: 1,
          name: '2B'
        }
      }
    }

    cartTree = [{
      id: 1,
      name: 'Section 1',
      subsections: [{
        name: 'Subsection 1',
        variables: [{
          id: 123,
          name: 'var123',
          label: 'var 123',
          subvariables: [{ id: 456 }, { id: 789 }],
          subsections: [1],
          subsection: 1
        }, {
          id: 456,
          name: 'var456',
          label: 'var 456',
          subvariable_of: { id: 123 },
          subvariables: [],
          subsections: [1, 2],
          subsection: 1
        }, {
          id: 789,
          name: 'var789',
          label: 'var 789',
          subvariable_of: { id: 123 },
          subvariables: [],
          subsections: [1],
          subsection: 1
        }]
      }]
    }, {
      id: 2,
      name: 'Section 2',
      subsections: [{
        name: 'Subsection 2',
        variables: [{
          id: 456,
          name: 'var456',
          label: 'var 456',
          subvariables: [],
          subsections: [2],
          subsection: 2
        }]
      }]
    }]

    store = new Vuex.Store({
      state,
      actions,
      mutations,
      getters
    })
  })

  it('renders cart view, showing the assesment labels sorted alphabetically', () => {
    // CartView types do not match, but they should
    // @ts-ignore
    const wrapper = mount(CartView, { stubs, store, localVue })
    expect(wrapper.findAll('li').at(0).text()).toEqual('var 123 ( 1A, 2C )')
    expect(wrapper.findAll('li').at(1).text()).toEqual('var 456 ( 2B )')
    expect(wrapper.find('#cart-view')).toMatchSnapshot()
  })

  it('renders a save button that saves the current state', async () => {
    actions.save.mockResolvedValue('12345')
    // CartView types do not match, but they should
    // @ts-ignore
    const wrapper = mount(CartView, { stubs, store, localVue, mocks })
    wrapper.find('.save').trigger('click')
    await flushPromises()
    expect(mocks.$router.push).toHaveBeenCalledWith({ name: 'load', params: { orderNumber: '12345' } })
  })

  it('has the first selection open on start', async () => {
    // @ts-ignore
    const wrapper = mount(CartView, { stubs, store, localVue, mocks })
    expect(wrapper.find('#accordion-0').isVisible()).toBeTruthy()
  })

  it('has a collapsable menu', async () => {
    // @ts-ignore
    const wrapper = mount(CartView, { stubs, store, localVue, mocks })
    const cartHeader = wrapper.find('.hoverable')
    cartHeader.trigger('click')
    expect(wrapper.find('#accordion-0').isVisible()).toBeFalsy()
  })

  it('has a collapsable menu', async () => {
    // @ts-ignore
    const wrapper = mount(CartView, { stubs, store, localVue, mocks })

    const cartHeader = wrapper.find('.hoverable')
    cartHeader.trigger('click')

    expect(wrapper.find('#accordion-0').isVisible()).toBeFalsy()
  })

  it('removes closed menu items', async () => {
    // @ts-ignore
    const wrapper = mount(CartView, { stubs, store, localVue, mocks })

    expect(wrapper.vm.openItems).toHaveLength(1)

    const cartHeader = wrapper.find('.hoverable')
    cartHeader.trigger('click')

    expect(wrapper.vm.openItems).toHaveLength(0)
  })

  it('adds open menu items', async () => {
    // @ts-ignore
    const wrapper = mount(CartView, { stubs, store, localVue, mocks })

    const cartHeader = wrapper.find('.hoverable')
    cartHeader.trigger('click')

    // the first element closes on click.
    expect(wrapper.vm.openItems).toHaveLength(0)

    cartHeader.trigger('click')
    // the element is openened
    expect(wrapper.vm.openItems).toHaveLength(1)
  })

  it('can render variables sets correctly', async () => {
    // @ts-ignore
    const wrapper = mount(CartView, { stubs, store, localVue, mocks })
    // @ts-ignore
    expect(wrapper.vm.variableSetClass(cartTree[0].subsections[0].variables[0])).toEqual('start')
    expect(wrapper.vm.variableSetClass(cartTree[0].subsections[0].variables[1])).toEqual('line')
    expect(wrapper.vm.variableSetClass(cartTree[0].subsections[0].variables[2])).toEqual('end')
  })
})
