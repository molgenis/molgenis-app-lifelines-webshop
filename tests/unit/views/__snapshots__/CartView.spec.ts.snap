// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartView.vue renders cart view, showing the assesment labels sorted alphabetically 1`] = `
<div id="cart-view">
  <div class="container-fluid">
    <div class="row">
      <div class="col-sm-12 col-md-12 col-lg-6">
        <h3 class="h4">lifelines-webshop-cart-header</h3>
        <toast-component-stub value="[object Object]" class="my-3"></toast-component-stub>
        <!---->
        <div role="tablist" class="mt-3 mb-2">
          <div class="mb-2 d-flex justify-content-end"><button type="button" class="btn btn-primary save">lifelines-webshop-save-btn-label</button> <button type="button" class="btn btn-success ml-2">lifelines-webshop-order-btn-label</button></div>
          <div class="card mb-2">
            <!---->
            <!---->
            <header role="tab" class="card-header pl-1 pt-2 pr-2 pb-2 hoverable">
              <h5 class="pl-3 m-0 text-black">
                Section 1
                <svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="caret-up" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512" class="collapse-tree-icon ml-2 svg-inline--fa fa-caret-up fa-w-10" style="transform: rotate(180deg); transition: transform 0.2s;">
                  <path fill="currentColor" d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z" class=""></path>
                </svg></h5>
            </header>
            <div id="accordion-0" class="collapse show" role="tabpanel">
              <div class="card-body mx-3 my-1">
                <!---->
                <!---->
                <div>
                  <h5 class="h6">Subsection 1</h5>
                  <ul>
                    <li class="subvariable-line start"><span>var 123 ( 1A, 2C )</span></li>
                    <li class="subvariable-line line"><span>var 456 ( 2B )</span></li>
                    <li class="subvariable-line end"><span>var 789 </span></li>
                  </ul>
                </div>
              </div>
            </div>
            <!---->
            <!---->
          </div>
          <div class="card mb-2">
            <!---->
            <!---->
            <header role="tab" class="card-header pl-1 pt-2 pr-2 pb-2 hoverable">
              <h5 class="pl-3 m-0 text-black">
                Section 2
                <svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="caret-up" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512" class="collapse-tree-icon ml-2 svg-inline--fa fa-caret-up fa-w-10" style="transform: rotate(0deg); transition: transform 0.2s;">
                  <path fill="currentColor" d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z" class=""></path>
                </svg></h5>
            </header>
            <div id="accordion-1" class="collapse" style="display: none;" role="tabpanel">
              <div class="card-body mx-3 my-1">
                <!---->
                <!---->
                <div>
                  <h5 class="h6">Subsection 2</h5>
                  <ul>
                    <li class="subvariable-line"><span>var 456 ( 2B )</span></li>
                  </ul>
                </div>
              </div>
            </div>
            <!---->
            <!---->
          </div>
        </div>
        <!---->
      </div>
    </div>
  </div>
</div>
`;
